cmd/kubeadm/app/phases/upgrade/health.go

```diff
@@ -67,7 +67,7 @@ func (c *healthCheck) Name() string {
 // - all control-plane Nodes are Ready
 // - (if static pod-hosted) that all required Static Pod manifests exist on disk
 func CheckClusterHealth(client clientset.Interface, cfg *kubeadmapi.ClusterConfiguration, ignoreChecksErrors sets.Set[string]) error {
-	fmt.Println("[upgrade] Running cluster health checks")
```

Comment:
The mixture of fmt, klog and printer in kubeadm code makes my brain hurt. It's so hard to understand the code and choose the right choice. 

IMO, making printer as input argument of all functions is not a good choice. However, I don't have a better idea to improve it.

Reply:
> The mixture of fmt, klog and printer in kubeadm code makes my brain hurt. It's so hard to understand the code and choose the right way.

it's definitely messy.

> IMO, making printer as input argument of all functions is not a good choice. However, I don't have a better idea to improve it.

passing printer makes sense to me, but let's try to keep this PR minimal like i mentioned earlier.

Reply:
After I build kubeadm with `WHAT=cmd/kubeadm make`, I tried to run `kubeadm upgrade plan` but got the following error:

```shell
root@test-control-plane:/# /kubeadm upgrade plan
[upgrade/config] Loading the kubeadm configuration
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
[preflight] Running pre-flight checks.
[upgrade] Fetching available versions to upgrade to
[upgrade/versions] Cluster version: v1.28.0
[upgrade/versions] FATAL: Couldn't parse kubeadm version: could not parse pre-release/metadata (-master+$Format:%H$) in version "v0.0.0-master+$Format:%H$"
To see the stack trace of this error execute with --v=5 or higher
```

@neolit123 Can you give me some guide on how to solve this problem?

Reply:
run with --v=5 and debug where the error is happening in the code.

Reply:
The version value is determined at compile time.

https://github.com/carlory/kubernetes/blob/07e59b013f0c9f366e20caf78e6cd9db9349c039/cmd/kubeadm/app/phases/upgrade/versiongetter.go#L91

```
root@test-control-plane:/# /kubeadm upgrade plan -v5
...
could not parse pre-release/metadata (-master+$Format:%H$) in version "v0.0.0-master+$Format:%H$"
Couldn't parse kubeadm version
k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade.(*KubeVersionGetter).KubeadmVersion
	k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade/versiongetter.go:95
k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade.GetAvailableUpgrades
	k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade/compute.go:90
k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade.runPlan
	k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade/plan.go:269
k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade.newCmdPlan.func1
	k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade/plan.go:74
github.com/spf13/cobra.(*Command).execute
	github.com/spf13/cobra@v1.7.0/command.go:940
github.com/spf13/cobra.(*Command).ExecuteC
	github.com/spf13/cobra@v1.7.0/command.go:1068
github.com/spf13/cobra.(*Command).Execute
	github.com/spf13/cobra@v1.7.0/command.go:992
k8s.io/kubernetes/cmd/kubeadm/app.Run
	k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:52
main.main
	k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
	runtime/proc.go:271
runtime.goexit
	runtime/asm_arm64.s:1222
[upgrade/versions] FATAL
k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade.runPlan
	k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade/plan.go:271
k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade.newCmdPlan.func1
	k8s.io/kubernetes/cmd/kubeadm/app/cmd/upgrade/plan.go:74
github.com/spf13/cobra.(*Command).execute
	github.com/spf13/cobra@v1.7.0/command.go:940
github.com/spf13/cobra.(*Command).ExecuteC
	github.com/spf13/cobra@v1.7.0/command.go:1068
github.com/spf13/cobra.(*Command).Execute
	github.com/spf13/cobra@v1.7.0/command.go:992
```

Reply:
> WHAT=cmd/kubeadm make

try `make all WHAT=cmd/kubeadm`
then the binary is found in _output

Reply:
`KUBE_GIT_VERSION=v1.30.0 WHAT=cmd/kubeadm make` works for me.